### PR TITLE
Some improvements to current enh functions

### PR DESCRIPTION
**Pull request by Emrys365**
This PR improves some existing enh functions:
 - STOI scores are now calculated as percentage to preserve higher precision. And ESTOI is added as a new metric.
 - The generated `spk?.scp` files for enhanced audios in stage 7 will now use absolute paths, so that other models/tasks can easily use the enhanced audios without worrying about path issues.
 - IRM and IBM can be now correctly calculated by adding a new argument `noise_spec`.
 - CIRM (complex ideal ratio mask) is added in `_create_mask_label`.
 - Some new loss functions are added: FrequencyDomainAbsCoherence, FrequencyDomainCrossEntropy, TimeDomainMSE, TimeDomainL1.
 - `--inference_enh_config` is added to [egs2/TEMPLATE/enh1/enh.sh](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/enh1/enh.sh). It allows changing some of the model attributes during inference. An example is shown below:
    <details><summary>expand to view content of `inference_beamformer.yaml`</summary><div>

    ```yaml
    encoder: stft
    encoder_conf:
        n_fft: 512
        hop_length: 256
        use_builtin_complex: True
    decoder: stft
    decoder_conf:
        n_fft: 512
        hop_length: 256
    separator: wpe_beamformer
    separator_conf:
        use_wpe: True
        use_dnn_mask_for_wpe: False
        use_beamformer: True
        ref_channel: 4
        use_noise_mask: True
        beamformer_type: mvdr
    ```

    </div></details>

     - Also added support for the newly introduced enh_s2t task in https://github.com/espnet/espnet/pull/4226
 - Add support for additional stats in loss calculation.
    > This is implemented by merging the dictionary `criterion.stats` (if it exists) into the main one.
 - Allow specifying the criterion name by users.

TODO (in a future PR):
  - [ ] <del>add support for calculating losses on specific model outputs instead of the default one (speech estimates).</del>

    > On second thought, I find it complicated for the model to know the corresponding label for a specified model output to calculate losses.
    > So this action needs better consideration.

### CODE REVIEW

**Comment by sw005320** [comment]
@simpleoier or @LiChenda,
Can I ask you to review this PR?

File: espnet2/enh/espnet_model.py [hunk 1]
```diff
@@ -160,10 +160,15 @@ def forward(
             elif isinstance(criterion, FrequencyDomainLoss):
                 # for the time-frequency domain criterions
                 if criterion.compute_on_mask:
-                    # compute on mask
+                    # compute loss on masks
+                    if noise_ref is not None:
+                        noise_spec = self.encoder(noise_ref.sum(1), speech_lengths)[0]
```
Comment (LiChenda) [inline espnet2/enh/espnet_model.py:165]: Better to update unit test to cover this line.

**Review by LiChenda — APPROVED** [review]
Good work! I just leave a minor comment. This PR is already in a good shape for me.

File: espnet2/enh/espnet_model.py (hunk 1 above)
Comment (Emrys365) [reply espnet2/enh/espnet_model.py:201]: I believe it has been covered by various test in [test/espnet2/enh/test_espnet_model.py](https://github.com/espnet/espnet/blob/master/test/espnet2/enh/test_espnet_model.py):
```python
test_forward_with_beamformer_net[loss_wrappers0-True-1-mask_mse-IBM-True]
test_forward_with_beamformer_net[loss_wrappers0-True-1-mask_mse-IBM-False]
test_forward_with_beamformer_net[loss_wrappers0-True-1-mask_mse-IRM-True]
test_forward_with_beamformer_net[loss_wrappers0-True-1-mask_mse-IRM-False]
...
test_forward_with_beamformer_net[loss_wrappers0-True-1-magnitude-IBM-True] 
test_forward_with_beamformer_net[loss_wrappers0-True-1-magnitude-IBM-False]
...
test_forward_with_beamformer_net[loss_wrappers0-True-1-spectrum-IBM-True]
test_forward_with_beamformer_net[loss_wrappers0-True-1-spectrum-IBM-False]
...
test_forward_with_beamformer_net[loss_wrappers0-True-1-spectrum_log-IBM-True]
test_forward_with_beamformer_net[loss_wrappers0-True-1-spectrum_log-IBM-False]
...
test_forward_with_beamformer_net[loss_wrappers0-True-2-mask_mse-IBM-True]
test_forward_with_beamformer_net[loss_wrappers0-True-2-mask_mse-IBM-False]
test_forward_with_beamformer_net[loss_wrappers0-True-2-mask_mse-IRM-True]
test_forward_with_beamformer_net[loss_wrappers0-True-2-mask_mse-IRM-False]
...
test_forward_with_beamformer_net[loss_wrappers0-True-2-magnitude-IBM-True]
test_forward_with_beamformer_net[loss_wrappers0-True-2-magnitude-IBM-False]
...
test_forward_with_beamformer_net[loss_wrappers0-True-2-spectrum-IBM-True]
test_forward_with_beamformer_net[loss_wrappers0-True-2-spectrum-IBM-False]
...
test_forward_with_beamformer_net[loss_wrappers0-False-3-spectrum_log-IBM-True]
test_forward_with_beamformer_net[loss_wrappers0-False-3-spectrum_log-IBM-False]
```

**Comment by Emrys365** [comment]
@simpleoier, can you check my modifications, especially for the egs2 part?

File: espnet2/tasks/enh.py [hunk 2]
```diff
@@ -99,8 +102,11 @@
         sdr=SDRLoss,
         ci_sdr=CISDRLoss,
         si_snr=SISNRLoss,
+        coh=FrequencyDomainAbsCoherence,
         mse=FrequencyDomainMSE,
```
Comment (simpleoier) [inline espnet2/tasks/enh.py:106]: I was thinking about to use mse_f and l1_f to be consistent with mse_t and l1_t. But it might break the pretrained models. I think it is fine.

File: espnet2/enh/loss/criterions/time_domain.py [hunk 3]
```diff
@@ -207,3 +209,65 @@ def forward(
...
+        return mseloss
+
+
+class TimeDomainL1(TimeDomainLoss):
+    def __init__(self):
+        super().__init__()
+
+    @property
+    def name(self) -> str:
+        return "time_L1_loss"
```
Comment (simpleoier) [inline espnet2/enh/loss/criterions/time_domain.py:251]: Not sure. Is it better to use time_domain_L1_loss or TD_L1_loss?

File: espnet2/enh/loss/criterions/time_domain.py [hunk 4]
```diff
@@ -207,3 +209,65 @@ def forward(
...
         return si_snr
+
+
+class TimeDomainMSE(TimeDomainLoss):
+    def __init__(self):
+        super().__init__()
+
+    @property
+    def name(self) -> str:
+        return "time_MSE_loss"
```
Comment (simpleoier) [inline espnet2/enh/loss/criterions/time_domain.py:220]: Not sure. Is it better to use time_domain_MSE_loss or TD_MSE_loss?

File: espnet2/tasks/enh.py (hunk 2 above)
Comment (Emrys365) [reply espnet2/tasks/enh.py:106]: Yes, it is better to use `mse_fd` and `mse_td`. I can also keep the old keys for backward compatibility.

File: espnet2/enh/loss/criterions/time_domain.py (hunk 3 above)
Comment (Emrys365) [reply espnet2/enh/loss/criterions/time_domain.py]: I think so. I will modify it. Thanks!

File: espnet2/enh/loss/criterions/time_domain.py (hunk 4 above)
Comment (Emrys365) [reply espnet2/enh/loss/criterions/time_domain.py]: ditto

**Comment by sw005320** [comment]
Thanks a lot, @Emrys365!
